modules/core/src/test/java/org/apache/ignite/internal/processors/cache/IgniteOOMWithoutNodeFailureAbstractTest.java

```diff
@@ -0,0 +1,223 @@
+            .setDataRegionName("tx");
+
+        cfg.setMetricsLogFrequency(1_000);
+        cfg.setFailureHandler(new StopNodeOrHaltFailureHandler(false, 0));
+        cfg.setCacheConfiguration(baseCfg, txCacheCfg);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        ignite = startGrids(1);
+
+        awaitPartitionMapExchange();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        super.afterTest();
+    }
+
+    /**
+     * Tests how eviction behaves when {@link DataRegionConfiguration#getEmptyPagesPoolSize()} is not enough for large
+     * values.
+     *
+     * @throws Exception If failed.
+     */
+    protected void testIgniteOOMWithoutNodeFailure(IgniteCache<Object, Object> cache, Consumer<Runnable> cacheOpWrapper) throws Exception {
+        List<IgniteInternalFuture<Object>> futures = new ArrayList<>();
+
+        Function<ThreadLocalRandom, byte[]> function = random -> new byte[random.nextInt(3) * 100 * 1024];
+
+        AtomicReference<Exception> exceptionsRef = new AtomicReference<>();
+
+        AtomicBoolean running = new AtomicBoolean();
+
+        IgniteInClosure<IgniteInClosure<ThreadLocalRandom>> task = op -> {
+            ThreadLocalRandom random = ThreadLocalRandom.current();
+
+            while (running.get()) {
+                try {
+                    cacheOpWrapper.accept(() -> op.apply(random));
+                }
+                catch (CachePartialUpdateException e) {
+                    exceptionsRef.compareAndSet(null, e);
+
+                    running.set(false);
+                }
+            }
+        };
+
+        for (int i = 0; i < 3; i++) {
+            IgniteInternalFuture<Object> fut = GridTestUtils.runAsync(() -> {
+                task.apply(random -> cache.put(random.nextInt(), function.apply(random)));
+            });
+
+            futures.add(fut);
+        }
+
+        for (int i = 0; i < 3; i++) {
+            IgniteInternalFuture<Object> fut = GridTestUtils.runAsync(() -> {
+                task.apply(random -> cache.replace(random.nextInt(), function.apply(random)));
+            });
+
+            futures.add(fut);
+        }
+
+        for (int i = 0; i < 3; i++) {
+            IgniteInternalFuture<Object> fut = GridTestUtils.runAsync(() -> task.apply(random -> {
+                Map<Integer, Object> map = new HashMap<>();
+
+                map.put(random.nextInt(), function.apply(random));
+                map.put(random.nextInt(), function.apply(random));
+
+                cache.putAll(map);
+            }));
+
+            futures.add(fut);
+        }
+
+        doSleep(30_000);
+
+        running.set(false);
+
+        for (IgniteInternalFuture<Object> future : futures)
+            future.get(5, TimeUnit.SECONDS);
+
+        assertNull(exceptionsRef.get());
```

Comment:
Maybe you still need to make sure that **DataRegion#emptyPagesPoolSize** has grown?